### PR TITLE
[boost-modular-build-helper] Fix issue that link incorrect dependent libraries

### PR DESCRIPTION
**Pull request by jgiannuzzi**
This PR is entirely based on PR #9224 by @PhoebeHui.
It has only been updated to resolve conflicts with master.

I opened it as a new PR as suggested by @JackBoosY [here](https://github.com/microsoft/vcpkg/pull/9224#issuecomment-621605097).

It fixes issues #7559, #8613.
It is related to issue #8165.

### CODE REVIEW

File: ports/boost-modular-build-helper/boost-modular-build.cmake [hunk 1]
```diff
@@ -81,6 +80,18 @@ function(boost_modular_build)
...
+            set(BOOST_LIB_SUFFIX ${BOOST_LIB_RELEASE_SUFFIX})
+            set(VARIANT "release")
+            set(BUILD_LIB_PATH "lib/")
+            configure_file(${_bm_DIR}/Jamroot.jam ${_bm_SOURCE_PATH}/Jamroot.jam @ONLY)
+        endif()
+        if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
+            set(BOOST_LIB_SUFFIX ${BOOST_LIB_DEBUG_SUFFIX})
+            set(VARIANT debug)
+            set(BUILD_LIB_PATH "debug/lib/")
+            configure_file(${_bm_DIR}/Jamroot.jam ${_bm_SOURCE_PATH}/Jamroot.jam @ONLY)
```
Comment (ras0219-msft) [inline ports/boost-modular-build-helper/boost-modular-build.cmake:93]: This overwrites the copy of Jamroot.jam produced above in the release case -- surely that can't be the intended behavior?

Comment (PhoebeHui) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:93]: Overwrites the copy of Jamroot.jam produced in the release case, this is intended behavior.

What I thought before: 
In current Jamroot.jam, it configured both release and debug libraries, but doesn’t distinguish them, when build release libraries, it try to find dependency port debug libraries, that cause build failures.

To fix this issue, configured the Jamroot.jam before building release libraries, and configured it with debug build info again when building debug libraries.

**Review by PhoebeHui — COMMENTED** [review]
@jgiannuzzi, thanks for the PR!

**Comment by FXTDwiki** [comment]
@jgiannuzzi  💯 thanks for the fix! It solved my issue like a charm!

File: ports/boost-modular-build-helper/boost-modular-build.cmake (hunk 1 above)
Comment (PhoebeHui) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:93]: Yes, overwrites the copy of Jamroot.jam produced above in the release case is intended behavior.

What I thought before:
In current Jamroot.jam, it configured both release and debug libraries, but doesn’t distinguish them, when build release libraries, it try to find dependency port debug libraries, that cause build failures.

To fix this issue, configured the Jamroot.jam before build release libraries, and configured it with debug build info again when build debug libraries.

**Review by PhoebeHui — APPROVED** [review]

**Comment by PhoebeHui** [comment]
@ras0219-msft, could you help futher review?

**Comment by josch** [comment]
While this pull request *does* fix the boost-fiber build failures when VCPKG_BUILD_TYPE is set to release, I am now unable to use FindBoost.cmake to successfully find boost if building it with this patch applied. Did you test using the resulting libraries with cmake?

File: ports/boost-modular-build-helper/boost-modular-build.cmake (hunk 1 above)
Comment (PhoebeHui) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:93]: Thanks Robert for pointing out this!

This is diferent with my original PR https://github.com/microsoft/vcpkg/pull/9224/, @jgiannuzzi, is there some reason to change the code here?

**Review by PhoebeHui — CHANGES_REQUESTED** [review]

File: ports/boost-modular-build-helper/boost-modular-build.cmake (hunk 1 above)
Comment (PhoebeHui) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:93]: I reopened PR #9224, and continue to follow up this issue.

Comment (jgiannuzzi) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:93]: Hi @PhoebeHui, I refactored your original code only to try to resolve conflicts with master with the best of my understanding of the build system. I'm very happy for you to take over and reopen the original PR.
All that matters to me is for the issue to be solved! 😄

Comment (jgiannuzzi) [reply ports/boost-modular-build-helper/boost-modular-build.cmake:93]: FWIW I think the refactoring you just did on the original PR makes more sense than what I did.

**Comment by PhoebeHui** [comment]
Duplicated this PR to https://github.com/microsoft/vcpkg/pull/9224. @jgiannuzzi, Appreciate your effort very much!

**Pull request closed without merging**